packages/discovery-react-components/src/components/DocumentPreview/components/PdfViewer/PdfViewerTextLayer.tsx

```diff
@@ -0,0 +1,207 @@
+import React, { FC, useEffect, useRef, useState } from 'react';
+import cx from 'classnames';
+import PDFJSLib, { PDFPageProxy, PDFPageViewport, TextContent, TextContentItem } from 'pdfjs-dist';
+import { EventBus } from 'pdfjs-dist/lib/web/ui_utils';
+import { TextLayerBuilder } from 'pdfjs-dist/lib/web/text_layer_builder';
+
+const { RenderingCancelledException } = PDFJSLib as any;
+
+interface Props {
+  className?: string;
+
+  /**
+   * PDF page from pdfjs
+   */
+  loadedPage: PDFPageProxy | null | undefined;
+
+  /**
+   * Page number, starting at 1
+   */
+  page: number;
+
+  /**
+   * Zoom factor, where `1` is equal to 100%
+   */
+  scale: number;
+
+  /**
+   * Callback for text layer info
+   */
+  setTextLayerInfo?: (info: PdfTextLayerInfo | null) => any;
+}
+
+export type PdfTextLayerInfo = {
+  /**
+   * PDF text content
+   */
+  textContent: TextContent & {
+    styles: { [styleName: string]: CSSStyleDeclaration };
+  };
+
+  /**
+   * Text span DOM elements rendered on the text layer
+   */
+  textDivs: HTMLElement[];
+
+  /**
+   * Pdf page viewport used to render text items
+   */
+  viewport: PDFPageViewport;
+
+  /**
+   * Page number, starting at 1
+   */
+  page: number;
+};
+
+const PdfViewerTextLayer: FC<Props> = ({
+  className,
+  loadedPage,
+  page = 1,
+  scale = 1,
+  setTextLayerInfo: setTextLayerInfoCallback = () => {}
+}) => {
+  const textLayerRef = useRef<HTMLDivElement>(null);
+  const textLayerDiv = textLayerRef.current;
+
+  const [textRenderInfo, setTextRenderInfo] = useState<{
+    page: number;
+    scale: number;
+    textContent: TextContent;
+    viewport: PDFPageViewport;
+  } | null>(null);
+
+  useEffect(() => {
+    async function loadTextInfo() {
+      const isPageReady = !!loadedPage && loadedPage.pageNumber === page;
+      if (isPageReady) {
+        const viewport = loadedPage.getViewport({ scale });
+        const textContent = await loadedPage.getTextContent();
+        setTextRenderInfo({ textContent, viewport, page, scale });
+      }
+    }
+    loadTextInfo();
+  }, [loadedPage, page, scale]);
+
+  const textLayerBuilderRef = useRef<TextLayerBuilder | null>(null); // ref for debugging purpose
+  const [textLayerInfo, setTextLayerInfo] = useState<PdfTextLayerInfo | null>(null);
+  useEffect(() => {
```

Comment:
This `useEffect` is quite large. I suggest refactoring this to a standalone function or even a hook (you can still have this within this same file; no reason to make a separate file for either approach).